### ISSUE TITLE
Support running dartanalyzer on the sdk root

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#35562]
Opening issue in sdk for https://github.com/Dart-Code/Dart-Code/issues/1295

It should be possible run the dartanalyzer on the root of this project. However, if run on the root it generates 10000+ errors:

> The SDK repository contains a lot of files that are not normal Dart but are used by the VM, dart2js and ddc to help implement the core libraries. It also contains a lot of test files that, by design, contain syntactic and semantic errors.

A solution would be to add an `analysis_options.yaml` file to the root:

```
analyzer:
  exclude:
    - docs/newsletter/20171103/**
    - out/**
    - runtime/**
    - samples-dev/swarm/**
    - sdk/lib/**
    - tests/**
    - third_party/observatory_pub_packages/**
    - third_party/pkg/**
    - third_party/pkg_tested/dart_style/**
    - third_party/tcmalloc/**
    - tools/apps/update_homebrew/**
    - tools/dart2js/**
    - tools/dom/**
    - tools/sdks/dart-sdk/lib/**
    - tools/status_clean.dart
    - xcodebuild/**
```

However, this breaks the analyzer bots:

```dart
FAILED: dart2analyzer-none release_x64 co19_2/LibTest/isolate/Isolate/spawn_A02_t05
Expected: Pass
Actual: MissingCompileTimeError
Missing expected compile error.

--- Command "dart2analyzer" (took 1ms):
DART_CONFIGURATION=ReleaseX64 out/ReleaseX64/dart-sdk/bin/dartanalyzer --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/cache/builder/sdk/.packages --format=machine --no-hints /b/s/w/ir/cache/builder/sdk/tests/co19_2/src/LibTest/isolate/Isolate/spawn_A02_t05.dart

exit code:
1

stderr:
No dart files found at: /b/s/w/ir/cache/builder/sdk/tests/co19_2/src/LibTest/isolate/Isolate/spawn_A02_t05.dart

--- Re-run this test:
python tools/test.py -m release -n analyzer-linux --write-logs co19_2/LibTest/isolate/Isolate/spawn_A02_t05
```

> @bwilkerson any reason adding analysis_options.yaml at the root would trigger No dart files found?

> I suspect that when analyzer is asked to analyze the test file it finds the analysis options file in the root directory, and that because the file excludes the test file the analyzer decides that there are no files to analyze. Seems like the command-line tool should analyze explicitly mentioned files even if they're excluded, but I'm guessing it doesn't work that way today.

> It should be fairly easy; it's just a question of when the work would be scheduled. A temporary solution might be best for now.

I landed a workaround in ea16f85f8c273876c59a935e33ebaa83945d9686.
Run `tools/generate_idefiles.py` to generate an `analysis_options.yaml` (which will be gitignored for now).

**Comment by dcharkes** [comment]
Now that `dart:ffi` has analyzer support, I cannot use the workaround anymore locally because I want to run the analyzer tests.

It would be really nice if we could get this sorted out.

**Comment by DanTup** [comment]
Is it only the test folder causing issues? Could all the other folders listed above be ignored in an analysis_options files committed to the repo?

And am I right in thinking that the reason analysis fails for the `test/` folder is because there is test code in there designed for tests that will generate errors? If so, is it possible that having two analysis_options files might solve it - eg. one at the root that ignores the folder, and another inside it that does not. When running the tests, wouldn't it walk up the tree and find the empty (non-excluding) one, but when analysing from the root, it'd be ignored?

**Comment by dcharkes** [comment]
@DanTup yes, that seems to be a workaround (see my edited post). However, I cannot simply generate that file in `./out/...` because those folders might not be there or deleted on purpose.

I think the conceptual issue is that we want a _different_ analysis options for the editor than for the testing framework. In the tests, the errors should surface, but in the IDE the errors should not surface (there would be 10's of thousands).

**Comment by DanTup** [comment]
Sorry, I missed that it was inside the `out` folder.

As a workaround for VS Code, you might be able to use [dart.analysisExcludeFolders](https://dartcode.org/docs/settings/#dartanalysisexcludedfolders) (set it at workspace level, and make it an array of relative paths, as in the exclude list above). I can't remember if you already tried that but I think it should work (though something that doesn't require each user to do the same would be better - and I suspect you won't want to commit .vscode/settings.json here).

**Comment by dcharkes** [comment]
Using `dart.analysisExcludedFolders` works. (Of course even when you open a file which is excluded, you do not see red squiggles.)

For other people looking for a workaround:

`.vscode/settings.json`

```json
{
    "dart.analysisExcludedFolders": [
        "docs/newsletter/20171103",
        "out",
        "runtime",
        "samples-dev/swarm",
        "sdk/lib",
        "sdk_nnbd/lib",
        "tests",
        "third_party/observatory_pub_packages",
        "third_party/pkg",
        "third_party/pkg_tested/dart_style",
        "third_party/tcmalloc",
        "tools/apps/update_homebrew",
        "tools/dart2js",
        "tools/dom",
        "tools/sdks/dart-sdk/lib",
        "tools/status_clean.dart",
        "xcodebuild",
    ]
}
```

**Comment by cskau-g** [comment]
> I suspect that when analyzer is asked to analyze the test file it finds the analysis options file in the root directory, and that because the file excludes the test file the analyzer decides that there are no files to analyze. Seems like the command-line tool should analyze explicitly mentioned files even if they're excluded, but I'm guessing it doesn't work that way today.

An additional +1 to that. I just spent an hour pulling my hair over analyzer insisting:

```
$ dart pkg/analyzer_cli/bin/analyzer.dart /.../out/.../test_file_that_definitely_exists.dart
No dart files found at: /.../out/.../test_file_that_definitely_exists.dart
```

Which in turn broke `dart tools/test.dart -n analyzer-asserts-strong-linux ...` locally.

I think there are a few separate issues here:
- Analyzer should not ignore a file if explicitly given by the user.
- And/Or analyzer should warn about ignoring files (or at least explain it better than claiming the file wasn't found).
- `tools/generate_idefiles.py` probably shouldn't ignore all of `out/**` as it breaks local testing.

**Comment by dcharkes** [comment]
With `dart.analysisExcludedFolders` working in DartCode, we could remove generating `analysis_options.yaml`.

For reference, the `settings.json` snippet that I'm currently using which enables commenting out individual folders when working on them: https://gist.github.com/dcharkes/66771d3726acd2a131f9ef29078ac92d

**Comment by srawlins** [comment]
CC @kallentu who worked in this area recently, I believe.

**Comment by eseidel** [comment]
I've posted https://dart-review.googlesource.com/c/sdk/+/303841.  Open to other solutions. 🤷   Discussion on https://github.com/dart-lang/sdk/issues/52380

**Comment by dcharkes** [comment]
I have a script locally in which I specify the folders that I'd like to be analyzed and then it generates the `dart.analysisExcludedFolders` for me, which I copy into my workspace ([gist](https://gist.github.com/dcharkes/0a8ca26776f68d79195cfc3666fa3be5)).  (@DanTup should we have a `analysisIncludedFolders`?)

The advantages of this is that I can allowlist certain subfolders in pkg, test, and tools that I want IDE support for (typically what I'm working on at that moment) and that I can rerun the script if new folders are added rather than having to hand ignore them.

I'm not very familiar with adding a workspace to the repo, but it's most likely not useful for people who regularly develop on the SDK. My workspace has a ton of settings ([gist](https://gist.github.com/dcharkes/4d0e72b7e996adb4653f899bb337d00b)). If the workspace in the folder can be ignored by users who already have their own workspace, I have no objections to adding a default one. (Though, I do expect people to accidentally start using that workspace, modifying it and diffs showing up in CLs.)

**Comment by DanTup** [comment]
> The advantages of this is that I can allowlist certain subfolders in pkg, test, and tools that I want IDE support for (typically what I'm working on at that moment) and that I can rerun the script if new folders are added rather than having to hand ignore them.

Is that really what we want? If new folders are excluded by default but should be included (to ensure you're not introducing errors in them) it would be easy to miss them. If new folders are included but should be excluded, it will be obvious from the errors and you'd have a prompt to ignore them.

That said, I can't think of a good reason why we _shouldn't_ support an explicit include list that overrides using the open folder. It would be _quite_ similar to a workspace, except with a workspace the settings need to be define in the workspace file, whereas using this include list would still allow settings to be in `.vscode/settings.json`.

> I'm not very familiar with adding a workspace to the repo, but it's most likely not useful for people who regularly develop on the SDK. My workspace has a ton of settings ([gist](https://gist.github.com/dcharkes/4d0e72b7e996adb4653f899bb337d00b)). If the workspace in the folder can be ignored by users who already have their own workspace, I have no objections to adding a default one.

AFAIK there's no such thing as a "default" workspace. The workspace is just a file that you have to open specifically. You could add 10 of them to the SDK root folder, but they wouldn't do anything unless you explicitly opened them. Opening the root _folder_ itself would behave exactly the same.

(However, if you have a single workspace file in a folder and open the folder, VS Code will say "Hey, there's a workspace file here, did you want to open that instead?")

**Comment by dcharkes** [comment]
> and you'd have a prompt to ignore them.

That's a manual step. I'm thinking users add things to the SDK all the time, including folders with tests which are expected to fail. So I have to do O(N) (SDK developers) times rerunning my script. If I add something I have to O(1) add that. Though, the script works for me. If no-one else would benefit from `analysisIncludedFolders` we might not add it.

> whereas using this include list would still allow settings to be in `.vscode/settings.json`.

We might not want to commit a `.vscode/settings.json`, since everyone might want a different set of include/exclude folders.

**Comment by DanTup** [comment]
> That's a manual step. I'm thinking users add things to the SDK all the time, including folders with tests which are expected to fail. So I have to do O(N) (SDK developers) times rerunning my script. If I add something I have to O(1) add that.

Oh sorry, are you talking about an opt-in list just for _you_? I had assumed you meant something that could be committed that was the full set of all folders that everyone would want (and that the split between new ignored/not-ignored folders isn't massively skewed in one direction or the other).

If you want something specifically for you, is there any advantaged of adding an include list versus just using a `.code-workspace` file? With ` .code-workspace` file (which you can store outside of the SDK folder), you can put settings and launch configurations in it, without having to worry about them being in `.vscode/` and being accidentally committed (plus you can have multiple "sets" of folders - for example I have an "Analysis Server.code-workspace" and a "DDS.code-workspace" that include different folders, launch configurations, and even settings (for ex. analyzer has sort-members-on-save but DDS does not).

**Comment by dcharkes** [comment]
> Oh sorry, are you talking about an opt-in list just for you? 

Yes, so that I don't have to use the script that converts the opt-in list to the exclude-list.

I also ignore a ton of folders not because they have failures, but because the analyzer takes a while to analyze the whole SDK.

**Comment by dcharkes** [comment]
I meant an include list inside the code-workspace.

The exclude-list is ~200 entries long: https://gist.github.com/dcharkes/4d0e72b7e996adb4653f899bb337d00b#file-dart-sdk-code-workspace-L441-L666
The include list is: ~30 lines long: https://gist.github.com/dcharkes/0a8ca26776f68d79195cfc3666fa3be5#file-generate_vscode_settings-dart-L4-L32

**Comment by DanTup** [comment]
I meant to list the folders as the workspace folders directly, here:

```
	"folders": [
		{
			"path": "sdk"
		}
	],
```

That way only they will be visible in the workspace which also means non-Analysis things like "Go to file" would also be excluding them.

**Comment by dcharkes** [comment]
I prefer to be able to see all the files in the whole SDK, I don't like switching workspaces. Often I don't know what other code I might need to touch or read upfront when working on something.

**Comment by DanTup** [comment]
Ah, got it. If you file an issue at https://github.com/Dart-Code/Dart-Code about supporting included folders, I can't think of anything off the top of my head that would prevent it (we'd have to merge them from multiple workspace folders but that doesn't seem like a problem).

**Comment by srawlins** [comment]
CC @eseidel @devoncarew @jacob314 I think this is working now? After a few recent commits.

**Comment by dcharkes** [comment]
> Ah, got it. If you file an issue at https://github.com/Dart-Code/Dart-Code about supporting included folders, I can't think of anything off the top of my head that would prevent it (we'd have to merge them from multiple workspace folders but that doesn't seem like a problem).

I've been working with a script that generates a list of excluded dirs from a list of included dirs. Pasting it here if it's useful for anyone: https://gist.github.com/dcharkes/be2f064ca5a21ea31f6c0c39304cc2e4  (Note it's tailored to my work, so you'd probably want to modify the list of directories.)

**Comment by davidmorgan** [comment]
I think this pretty much fixes it

https://dart-review.googlesource.com/c/sdk/+/357860

The problem is triggered by "lone files" that are not part of a package, for example most language tests.

What's happening is that the LSP server finds their only enclosing context as the SDK root--because that's where the .dart_tool folder is. It then adds the SDK root as somewhere to look for packages, exactly as if you had created a workspace containing the whole SDK.

A simple fix, although I do not know if it is correct, is to only treat a discovered enclosing context as a new workspace root if it is itself a package ("contains a lib folder").

Now you can just use "add folder to workspace" to open whatever subset of the SDK you want, and it will resolve that and deps and nothing else.

I hope :)

No idea if this is landable--Danny, what do you think please?

**Comment by DanTup** [comment]
I added some notes on the review (tl;dr - I don't understand how this would fix it, but if you can confirm it does for you and how you're opening the folder/etc., I could do some testing).

> Now you can just use "add folder to workspace" to open whatever subset of the SDK you want, and it will resolve that and deps and nothing else.

FWIW, my understanding was that this has always worked (I work on several packages in the SDK by just adding their folders to my workspace). I thought this issue was specifically about opening the _entire_ SDK where you'd get a lot of "invalid" code also in the workspace being analyzed. If you selectively pick just the bits you're interested in, I would things to just work (although, I'll note that I'm always working inside `pkg/` and therefore I'm not entirely sure the rules for other bits of the SDK outside of that behave the same).

**Comment by davidmorgan** [comment]
Thanks Danny :) replied on the review with more detail.

I suspect opening files/folders under `pkg` probably does mostly work, as they are packages, so you might not ever hit any files that don't have a package.

It's possible my use case is not quite what the rest of this issue is asking for :) but I think it mostly boils down to the same thing, a way to vary the set of files you're interested in under the SDK root and then analyze only those files.

**Comment by dcharkes** [comment]
> I think this pretty much fixes it
> 
> https://dart-review.googlesource.com/c/sdk/+/357860
> 
> The problem is triggered by "lone files" that are not part of a package, for example most language tests.
> 
> What's happening is that the LSP server finds their only enclosing context as the SDK root--because that's where the .dart_tool folder is. It then adds the SDK root as somewhere to look for packages, exactly as if you had created a workspace containing the whole SDK.

I am definitely creating the workspace as the whole SDK, I don't add individual directories in the SDK to a workspace.

Does that mean what you are proposing is a no-op for my workflow?

**Comment by dcharkes** [comment]
I actually like the whole SDK to be analyzed. That way if I update some DEPS file, pull in a package that has a breaking change to third_party/pkg and that is used somewhere else in the SDK I get a list of places I need to update. So in some sense explicitly opting out the directories fits with my workflow well.

End yes, I want to browse the whole file tree. My brain works best with predictable locations.

Anyway, let me not be the one preventing people from optimizing the work flow where they select individual dirs to opt in to their workspace. 😄 

I just wanted to understand the solution you're proposing.

**Comment by davidmorgan** [comment]
Makes sense, thanks!

Hopefully knowing that "onlyAnalyzeProjectsWithOpenFiles=true" can currently pull in the whole SDK when you didn't mean to is a small contribution to the overall puzzle here ;)

**Comment by DanTup** [comment]
> Hopefully knowing that "onlyAnalyzeProjectsWithOpenFiles=true" can currently pull in the whole SDK when you didn't mean to is a small contribution to the overall puzzle here ;)

If I understand correctly, this only happens if your open files include something outside of a "project"? (in which case, I wonder what the expectation here should be.. if those things are "projects" perhaps we can make the server understand where their roots really are.. if they're not, then I'm not sure whether it's entirely valid to analyze them as loose files.

Going back to the original problem.. if I understand correctly it is this:

- The SDK contains a bunch of "invalid" code (for tests, patch files, etc.) that we want to exclude if you open the SDK root
- These things can't be listed in analysis_options as excludes, because it breaks bots that try to analyze so of them

If this is the case and we can focus only on VS Code, then I wonder why we can't just committ a `.vscode/settings.json` into the SDK root with the necessary exclusions? (while there are some comments above about an include list being better, I think this could also result in new code that should be analyzed being excluded, and it may be better to default to including and adding exclusions being a manual step?).

**Comment by davidmorgan** [comment]
Those "loose" files can be analyzed just fine using the "whole SDK" package config, and that works today.

The unexpected part is that having found the "whole SDK" config it goes and proactively reads the whole SDK :) ... okay, I guess technically this is "only analyze projects with open files", and because of how we have it set up with a single package config at the root, the "project" is the whole SDK. I guess it is the word "only" that is misleading--just for our case. For our case it would be better named "proactivelyAnalyzeWholeProjectsWithOpenFiles" :)

Maybe what I'm asking for is a different feature altogether, but it strikes me--_why_ are we proactively analyzing files? That seems like more a thing to do in the "specified workspace folders" case than the "whatever files are opened" case. We could I guess have a mode where we find package config based on open files, but never proactively analyze.

If you think this is worth discussing then maybe let's fork an issue for it, as you say it's a different problem? :)

Re: the actual problem being discussed :) ... my understanding is that it's not as simple as excluding files, you do sometimes want those paths to be analyzed--if you are working on an "error test", for example, you want to analyze those broken files. So I guess the "whole SDK" question is really about the user experience of toggling analysis for parts of the tree.

Is there any possibility of making the explorer view directly support the concept of "Dart analysis is enabled here"? So you can see in the tree what is currently analyzed, and right click to toggle it on/off? That would be awesome :) ... another power user feature would be to right click on a file in explorer and say "analyze all my deps", then that would turn on analysis for all deps that are "checked out", by which I mean the ones that are resolved via deps overrides instead of served from pub. You could also have "analyze just my deps" which simultaneously turns off analysis for everything else in the explorer view.

**Comment by dcharkes** [comment]
> Maybe what I'm asking for is a different feature altogether, but it strikes me--_why_ are we proactively analyzing files?

You change some API in a file that is used in other files, you want errors to show up?

> Re: the actual problem being discussed :) ... my understanding is that it's not as simple as excluding files, you do sometimes want those paths to be analyzed--if you are working on an "error test", for example, you want to analyze those broken files. 

Correct, I then go and un-ignore the dir that file is in, work on it, and re-ignore it after. I'm not sure if there could be a better workflow.

> So I guess the "whole SDK" question is really about the user experience of toggling analysis for parts of the tree.

Yeah, so in effect I'm doing that toggling through modifying `dart.analysisExcludedFolders`.

> Is there any possibility of making the explorer view directly support the concept of "Dart analysis is enabled here"? So you can see in the tree what is currently analyzed, and right click to toggle it on/off? That would be awesome :)

That would be more convenient indeed that editing the workspace JSON! 😄  (It should probably have the effect of editing the workspace JSON for `analysisExcludedFolders`.)

**Comment by davidmorgan** [comment]
> > Maybe what I'm asking for is a different feature altogether, but it strikes me--_why_ are we proactively analyzing files?
> 
> You change some API in a file that is used in other files, you want errors to show up?

Oh, yeah, that's the reason. Sorry, I'm so used to working in google3 where that's impossible for the IDE to do that I keep forgetting ;) thanks.

I guess the power user feature for this would maybe be "analyze exactly the rdeps of my open/edited files" :)

**Comment by DanTup** [comment]
Yeah, the "normal" behaviour here is to analyze your _workspace_. When a user opens a Dart project they typically want the Problems view to populate with all errors for the project, not only their open files.

The problem that came up was that some terminal editors would set the "workspace root" as the home directory or a directory where you opened it from, so users might type:

```
my-editor projects/foo/lib/foo.dart
```

Wanting to just edit that single file, but the LSP server was told the home directory had been opened and therefore would analyze the whole thing. So we added `onlyAnalyzeProjectsWithOpenFiles` as a way to say "just ignore my open workspace folders, and instead look at the open file(s) and figure out here their project root is". We still want to analyze the whole project here (so if you're making changes in that file that cause errors, you can see them).

The issue here is that the mechanism for knowing what is a "project" doesn't work for many folders here in the SDK. We could implement some "onlyAnalyzeOpenFiles` flag that strictly analyzes only the open files, but I'm not sure if all the cases here really are loose files or if some of them are actually "projects" (for example, are there folders listed in the `package_config` but they don't have their own pubspecs?).




> Is there any possibility of making the explorer view directly support the concept of "Dart analysis is enabled here"? So you can see in the tree what is currently analyzed, and right click to toggle it on/off? That would be awesome :) 

I don't know what "see" means here - we have limited ability to customise this tree.. we can add context menu entries to exclude from analysis (although I don't think we could easily indicate whether they were already excluded or not because the context menu is somewhat statically declared).

**Comment by DanTup** [comment]
We can add badges to explorer like this:

![image](https://github.com/dart-lang/sdk/assets/1078012/20578cec-61fe-4a57-b8e5-804693d1c283)

(and as odd as it seems, despite this, we cannot conditionally show the context menu entry for Include/Exclude based on this... VS Code will asynchronously ask us for badgers, but it will not asynchronously wait for context menu so those conditions need to be static)

So, while not perfect we could potentially:

1. Show an "ex" badge in explorer on folders explicitly excluded by the workspaces `analysisExcludedFolders` setting
2. Have an "Exclude folder from Dart analysis" context menu that adds the folder to `analysisExcludedFolders` for the workspace
3. Have an "Include folder in Dart analysis" context menu that removes the folder to `analysisExcludedFolders` for the workspace (but it would show for all folders, even if not already excluded)

Or we could swap 2/3 for a single entry that just toggles the current behaviour (and rely on the badge to indicate if excluded)

I don't know if this would solve all of the problems though (unless a default ignore set can be agreed and committed.. and then probably gitignored to avoid people modifying it?).

**Comment by dcharkes** [comment]
> So, while not perfect we could potentially:
> 
> 1. Show an "ex" badge in explorer on folders explicitly excluded by the workspaces `analysisExcludedFolders` setting
> 2. Have an "Exclude folder from Dart analysis" context menu that adds the folder to `analysisExcludedFolders` for the workspace
> 3. Have an "Include folder in Dart analysis" context menu that removes the folder to `analysisExcludedFolders` for the workspace (but it would show for all folders, even if not already excluded)
> 
> Or we could swap 2/3 for a single entry that just toggles the current behaviour (and rely on the badge to indicate if excluded)

Either one works for me!

> I don't know if this would solve all of the problems though (unless a default ignore set can be agreed and committed.. and then probably gitignored to avoid people modifying it?).

With my experience, I don't think we can agree on a default. I'm just one engineer and already toggling. For example, I regularly toggle `runtime/tests/vm/dart` because it has about ~1000 errors/warnings, and every once in a while I need to touch tests there.

**Comment by davidmorgan** [comment]
Thanks Danny, the badges + context menu look great!

How could it work with subfolders? E.g. if "foo" is excluded:

```
foo <ex>
```

then you expand it

```
foo/bar 
foo/baz
foo/bop
```

do `bar` and `baz` show as excluded? What does the context menu do?

I'm thinking about if you want to un-exclude `baz`, then the config change is quite tricky: you have to un-exclude `foo` then exclude all subfolders except `baz`:

```
foo/bar <ex>
foo/baz
foo/bop <ex>
```

which is going to lead to a big+fragile config. (Which is a bit why we have this discussion in the first place).

Off the top of my head I guess the most robust config is to have both an include and an exclude list, and then whichever matches closest wins. So you can exclude `foo` then include `foo/baz` and it works as you want.

Hmmmm also off the top of my head, the context menu would then have three options :/ ... "exclude", "include" and "inherit". That's a bit horrible, but I can't come up something better :) what do you think please?

Thanks!

**Comment by DanTup** [comment]
> do `bar` and `baz` show as excluded? What does the context menu do?

I was thinking that for simplicity we should probably only include the badge on the folders explicitly included in `dart.analysisExcludedFolders`. If we have a single toggle on the context menu it'll then be clear exactly what it'll do (either add or remove that folder), and if we have individual commands it won't be confusing if you're trying to "unexclude" a sub-folder.


> Hmmmm also off the top of my head, the context menu would then have three options :/ ... "exclude", "include" and "inherit". That's a bit horrible, but I can't come up something better :) what do you think please?

Yeah, I don't think we should try to build something complex here (especially given how limited our customisation of the explorer is). This feature is already a bit of a duplicate of analysis_options exclude options but we kept it because it allows excluding non-projects, like `node_modules` which can be useful for performance.

I think making it easier to use/see the current settings is a good idea, but I'm less sure that we should be building more functionality for inclusions (un-excludes) and inheriting directly into VS Code (/ a non-standard LSP setting) in this way.

I don't know if we can handle context menu commands for multi-select, but if we could maybe excluding all sub-folders wouldn't be lots of clicks? It'd need some testing though.

**Comment by DanTup** [comment]
Related: I opened https://github.com/microsoft/vscode/issues/208079 about the idea of allowing extensions to attach contexts to explorer tree nodes (which would allow showing the "correct" context menu command based on the current exclude status).

**Comment by davidmorgan** [comment]
Thanks Danny!

This is a hard problem ;)

Here is a realistic example without un-excludes: I might want to go from completely excluding `tests/` (I mean the top level SDK folder `dart-sdk/sdk/tests`, not anything in a package) to analyzing only its subfolder `tests/language/macros/augment`.

This requires excluding 10 folders under `tests`, then 134 folders under `tests/language`, then 4 folders under `tests/language/macros`.

Then reverse that to go back to the previous state when you're done.

The "reverse" part is more attractive to automate: if you click "exclude" on `tests/` we could remove all excludes under that path. That leads you with a nice clean configuration--if it's what you want! But it's expensive to undo if you want to go back to the fine grained config for `tests/language/macros/augment`.

If we automate both, so you can click "include" on `tests/language/macros/augment` and it adds 10+134+4 excludes for you, that works as long as nothing changes, but does not do what you want when a new folder is added e.g. under `tests/language`.

And then with multiple such "inclusions" in flight I think it quickly gets unmanageable :/ ... well, it is manageable with a script in which you list the includes and can regenerate the config ;) which I think is about where the issue got to so far.

**Comment by dcharkes** [comment]
> And then with multiple such "inclusions" in flight I think it quickly gets unmanageable :/ ... well, it is manageable with a script in which you list the includes and can regenerate the config ;) which I think is about where the issue got to so far.

👍  Right, that's why I have the script, it generates the 10+134+4 excludes.

**Comment by DanTup** [comment]
The VS Code bot is threatening to close https://github.com/microsoft/vscode/issues/208079 because it doesn't have 20 👍's.. if you want to see something like being able to easily include/exclude from the explorer menu, please add your 👍 to that issue!

**Comment by davidmorgan** [comment]
Thanks Danny :)
